modules/spring-cache-ext/src/main/java/org/apache/ignite/cache/spring/IgniteClientSpringCacheManager.java

```diff
@@ -0,0 +1,216 @@
+ *            http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans.xsd
+ *            http://www.springframework.org/schema/cache http://www.springframework.org/schema/cache/spring-cache.xsd"&gt;
+ *     &lt;!-- Provide Ignite client instance. --&gt;
+ *     &lt;bean id="cacheManager" class="org.apache.ignite.cache.spring.IgniteClientSpringCacheManager"&gt;
+ *         &lt;property name="clientConfiguration"&gt;
+ *             &lt;bean class="org.apache.ignite.configuration.ClientConfiguration"&gt;
+ *             &lt;property name="addresses"&gt;
+ *                 &lt;list&gt;
+ *                     &lt;value&gt;127.0.0.1:10800&lt;/value&gt;
+ *                 &lt;/list&gt;
+ *             &lt;/property&gt;
+ *         &lt;/bean&gt;
+ *         &lt;/property&gt;
+ *     &lt;/bean>
+ *
+ *     &lt;!-- Use annotation-driven cache configuration. --&gt;
+ *     &lt;cache:annotation-driven/&gt;
+ * &lt;/beans&gt;
+ * </pre>
+ */
+public class IgniteClientSpringCacheManager extends AbstractCacheManager implements DisposableBean,
+    ApplicationListener<ContextRefreshedEvent>
+{
+    /** Ignite client instance. */
+    private IgniteClient cli;
+
+    /** Ignite client configuration. */
+    private ClientConfiguration cliCfg;
+
+    /** Cache configurations mapped to cache name. */
+    private Map<String, ClientCacheConfiguration> ccfgs;
+
+    /** Flag that indicates whether Ignite client instance was set explicitly. */
+    private boolean externalCliInstance;
+
+    /** Gets Ignite client instance. */
+    public IgniteClient getClientInstance() {
+        return cli;
+    }
+
+    /** Sets Ignite client instance. */
+    public IgniteClientSpringCacheManager setClientInstance(IgniteClient cli) {
+        A.notNull(cli, "cli");
+
+        this.cli = cli;
+
+        externalCliInstance = true;
+
+        return this;
+    }
+
+    /** Gets Ignite client configuration. */
+    public ClientConfiguration getClientConfiguration() {
+        return cliCfg;
+    }
+
+    /** Sets Ignite client configuration that will be used to start the client instance by the manager. */
+    public IgniteClientSpringCacheManager setClientConfiguration(ClientConfiguration cliCfg) {
+        this.cliCfg = cliCfg;
+
+        return this;
+    }
+
+    /** Gets Ignite cache configurations. */
+    public Collection<ClientCacheConfiguration> getCacheConfigurations() {
+        return ccfgs == null ? Collections.emptyList() : ccfgs.values();
+    }
+
+    /**
+     * Sets the Ignite cache configurations that will be used to start the Ignite cache with the name requested by
+     * the Spring Cache Framework if one does not exist. If configuration with the requested name
+     * is not found among the configurations specified by this method, the default configuration with the requested
+     * name will be used.
+     */
+    public IgniteClientSpringCacheManager setCacheConfigurations(ClientCacheConfiguration... cfgs) {
```

Comment:
Why are we use different approaches for thin and thick clients? Can we use one dynamic configuration and create new caches based on this configuration changing only the cache name?

Reply:
At first I was stuck with ClientCacheConfiguration not implementing copy constructor as CacheConfiguration does.
Then I thought it would be quite useful to be able to provide a unique configuration for each cache, rather than one for all. I think it's not a problem to add dynamic configuration approach for IgniteClientSpringCacheManager but the current approach includes it. WDYT?

Reply:
Unique configuration for each cache can be defined in Ignite node configuration.
The current approach doesn't include a dynamic cache configuration approach, you should always set cache configuration explicitly for each cache.

Reply:
Done. Fresh TC build run - https://ci.ignite.apache.org/project.html?projectId=IgniteExtensions&branch_IgniteExtensions=pull%2F46%2F

Reply:
I'm sorry, I spoke too soon. It's not finished yet.

Reply:
Done.